hibernate-core/src/main/java/org/hibernate/query/sqm/internal/SqmUtil.java

```diff
@@ -827,10 +827,19 @@ static JpaOrder sortSpecification(SqmSelectStatement<?> sqm, Order<?> order) {
 	}
 
 	public static boolean isSelectionAssignableToResultType(SqmSelection<?> selection, Class<?> expectedResultType) {
-		if ( expectedResultType == null
-				|| selection != null && selection.getSelectableNode() instanceof SqmParameter ) {
+		if ( expectedResultType == null ) {
 			return true;
 		}
+		else if ( selection != null && selection.getSelectableNode() instanceof SqmParameter ) {
+			// assume the parameter type is assignable to the result type
+			if ( expectedResultType.isArray() ) {
+				final SqmParameter<?> parameter = (SqmParameter<?>) selection.getSelectableNode();
+				return parameter.allowMultiValuedBinding();
```

Comment:
Without this change, this method was returning true for any parameter type - this means that a query like `session.createQuery( "select ?1 from dual", Obect[].class ).setParameter( 1, 1 )` was returning `1` (since we were selecting `RowTransformerSingularReturnImpl`) instead of `Object[] { 1 }`, which is the requested result type.

The idea was that if a parameter allows multi-valued binding, then the unboxing might be desirable to avoid getting a nested array as a result type, though to ensure the parameter is indeed multi-valued we would need access to the bindings that don't exist yet at this time.

I guess we might as well always return `false` here for parameters and we would comply with the requested return type, wdyt?

Reply:
Yes, always returning `false` sounds sensible.